utils_cv/detection/model.py

```diff
@@ -285,63 +380,138 @@ def plot_precision_loss_curves(
 
-        ax1.set_xlim([0, self.epochs - 1])
-        ax1.set_xticks(range(0, self.epochs))
-        ax1.set_title("Loss and Average Precision over epochs")
-        ax1.set_xlabel("epochs")
-        ax1.set_ylabel("loss", color="g")
-        ax1.plot(self.losses, "g-")
+        for i, (k, v) in enumerate(ap.items()):
 
-        ax2 = ax1.twinx()
-        ax2.set_ylabel("average precision", color="b")
-        ax2.plot(self.ap, "b-")
+            ax1 = fig.add_subplot(1, len(ap), i+1)
 
-    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
+            ax1.set_xlim([0, self.epochs - 1])
+            ax1.set_xticks(range(0, self.epochs))
+            ax1.set_title("Loss and Average Precision over epochs")
+            ax1.set_xlabel("epochs")
+            ax1.set_ylabel("loss", color="g")
+            ax1.plot(self.losses, "g-")
+
+            ax2 = ax1.twinx()
+            ax2.set_ylabel("average precision for ".format(k), color="b")
+            ax2.plot(v, "b-")
+
+    def evaluate(self, dl: DataLoader = None) -> Union[CocoEvaluator, None]:
         """ eval code on validation/test set and saves the evaluation results in self.results. """
         if dl is None:
+            if not self.dataset:
+                return
             dl = self.dataset.test_dl
         self.results = evaluate(self.model, dl, device=self.device)
         return self.results
 
+    def _transform(
+        self,
+        im: Union[str, Path, Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+        """ Convert the image to the format required by the model. """
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im)
+        if self.device:
+            im = im.to(self.device)
+        return im
+
+    @classmethod
+    def _apply_threshold(
+        cls,
+        pred: Dict,
+        threshold: Optional[int] = 0.5,
+        mask_threshold: Optional[int] = 0.5,
+        **kwargs,
+    ) -> Dict:
+        """ Return prediction results that are above the threshold if any. """
+        # detach prediction results to cpu
+        pred = {k: v.detach().cpu().numpy() for k, v in pred.items()}
+        # apply score threshold
+        if threshold:
+            selected = pred['scores'] > threshold
+            pred = {k: v[selected] for k, v in pred.items()}
+        if "masks" in pred and mask_threshold:
+            pred["masks"] = pred["masks"] > mask_threshold
+        return pred
+
+    def _get_det_bboxes(
+        self,
+        pred: Dict,
+        im_path: Union[str, Path]
+    ) -> List[DetectionBbox]:
+        """ Gets the bounding boxes and labels from the prediction object. """
+        return DetectionBbox.from_arrays(
+            pred['boxes'].tolist(),
+            score=pred['scores'].tolist(),
+            label_idx=pred['labels'].tolist(),
+            label_name=np.array(self.labels)[pred['labels'] - 1].tolist(),
+            im_path=im_path,
+        )
+
+    def _process_pred_results(
```

Comment:
this function seems extraneous. Instead of `_get_det_bboxes`, maybe we can expand the function to `_get_det_bboxes_and_mask`, and then we won't need this function anymore. i think its also semantically a bit clearer

Reply:
:+1:  Good point!  Thanks!  Done

Reply:
:+1:  Good point!  Thanks!  Done